pandas/core/reshape.py

```diff
@@ -558,87 +549,138 @@ def _convert_level_number(level_num, columns):
+    new_levels += level_vals
+    new_labels += [np.tile(labels, N) for labels in zip(*level_labels)]
+    new_names += level_vals_used.names
+    new_index = MultiIndex(levels=new_levels, labels=new_labels,
+                           names=new_names, verify_integrity=False)
+
+    # if stacking all levels in columns, result will be a Series
+    if len(frame.columns.levels) == num_levels_to_stack:
+        new_data = frame.values.ravel()
+        if dropna:
+            mask = notnull(new_data)
+            new_data = new_data[mask]
+            new_index = new_index[mask]
+        return Series(new_data, index=new_index)
+
+    # result will be a DataFrame
+
+    def _equal_key(keya, keyb):
+        return all(((a == b) or
+                    ((isinstance(a, float) and np.isnan(a) and isinstance(b, float) and np.isnan(b)) or
+                     (isinstance(a, NaTType) and isinstance(b, NaTType))))
+                   for a, b in zip(keya, keyb))
+
+    # construct new_columns
+    if len(frame.columns.levels) > (num_levels_to_stack + 1):
+        # result columns will be a MultiIndex
+        # tuple list excluding level for grouping columns
         tuples = list(zip(*[
-            lev.take(lab) for lev, lab in
-            zip(this.columns.levels[:-1], this.columns.labels[:-1])
+            _make_new_index(lev, lab) for lev, lab in
+            zip(this.columns.levels[:-num_levels_to_stack],
+                this.columns.labels[:-num_levels_to_stack])
         ]))
-        unique_groups = [key for key, _ in itertools.groupby(tuples)]
-        new_names = this.columns.names[:-1]
+        unique_groups = [x for i, x in enumerate(tuples) if (i == 0) or not _equal_key(x, tuples[i - 1])]
+        new_names = this.columns.names[:-num_levels_to_stack]
         new_columns = MultiIndex.from_tuples(unique_groups, names=new_names)
     else:
-        new_columns = unique_groups = this.columns.levels[0]
+        # result columns will be an Index
+        new_columns = _make_new_index(this.columns.levels[0], sorted(set(this.columns.labels[0])))
+        unique_groups = list(new_columns)
 
-    # time to ravel the values
+    # construct new_data
     new_data = {}
-    level_vals = this.columns.levels[-1]
-    level_labels = sorted(set(this.columns.labels[-1]))
-    level_vals_used = level_vals[level_labels]
-    levsize = len(level_labels)
     drop_cols = []
-    for key in unique_groups:
-        loc = this.columns.get_loc(key)
-        slice_len = loc.stop - loc.start
+    try:
+        nan_index = level_vals_used.get_loc(np.nan)
+        if not isinstance(nan_index, int):
+            nan_index = None
+    except KeyError:
+        nan_index = None
+    unique_group_levels = this.columns.nlevels - num_levels_to_stack
+    columns_values = this.columns.get_values()
+    for i, key in enumerate(unique_groups):
+        # The following nonsense is needed to properly handle Nans. Otherwise would simply do:
```

Comment:
There's a couple things going on.

First, ``MultiIndex.get_loc()`` doesn't work for a tuple with ``NaN``. That's why I do the ``_equal_key()`` thing.
```
In [4]: mi1 = MultiIndex.from_tuples([(11,'A'),(22,'B')])

In [5]: mi1
Out[5]:
MultiIndex(levels=[[11, 22], ['A', 'B']],
           labels=[[0, 1], [0, 1]])

In [6]: mi1.get_loc((22,))
Out[6]: slice(1, 2, None)

In [8]: mi2 = MultiIndex.from_tuples([(11,'A'),(np.nan,'B')])

In [9]: mi2
Out[9]:
MultiIndex(levels=[[11], ['A', 'B']],
           labels=[[0, -1], [0, 1]])

In [10]: mi2.get_loc((np.nan,))
C:\Python34\lib\site-packages\pandas\core\index.py:4159: PerformanceWarning: indexing past lexsort depth may impact performance.
  PerformanceWarning)
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-10-1a49e94036c2> in <module>()
----> 1 mi2.get_loc((np.nan,))

C:\Python34\lib\site-packages\pandas\core\index.py in get_loc(self, key)
   4162
   4163         for i, k in enumerate(follow_key, len(lead_key)):
-> 4164             mask = self.labels[i][loc] == self.levels[i].get_loc(k)
   4165             if not mask.all():
   4166                 loc = loc[mask]

C:\Python34\lib\site-packages\pandas\core\index.py in get_loc(self, key)
   1400         loc : int if unique index, possibly slice or mask if not
   1401         """
-> 1402         return self._engine.get_loc(_values_from_object(key))
   1403
   1404     def get_value(self, series, key):

pandas\index.pyx in pandas.index.IndexEngine.get_loc (pandas\index.c:3807)()

pandas\index.pyx in pandas.index.IndexEngine.get_loc (pandas\index.c:3648)()

pandas\index.pyx in pandas.index.Int64Engine._check_type (pandas\index.c:7299)()

KeyError: nan
```

Second, ``DataFrame.reindex()`` doesn't properly handle columns with ``NaN``s in them, so I need to copy those columns separately. That's what the ``nan_index`` is for.